File: cc2.4/cc_2_4.py
#! /usr/bin/env python

def rotateMatrix(matrix):
    n = len(matrix)
    for i in range(0, n/2):
        for j in range(0, n - 1 - 2*i):
            k = i+j
            tmp = matrix[i][k]
            matrix[i][k] = matrix[n-1-k][i]
            matrix[n-1-k][i] = matrix[n-1-i][n-1-k]
            matrix[n-1-i][n-1-k] = matrix[k][n-1-i]
            matrix[k][n-1-i] = tmp

def printMatrix(matrix):
    n = len(matrix)
    if n == 0:
        return
    m = len(matrix[0])
    outstr = ''
    for i in range(0, n):
        for j in range(0, m):
            outstr += ('%d ' % matrix[i][j])
        outstr += '\n'
    print outstr

def main():
    m1 = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    m2 = [[1, 2, 3, 4],
          [5, 6, 7, 8],
          [9, 0, 1, 2],
          [3, 4, 5, 6]]
    printMatrix(m2)
    rotateMatrix(m2)
    printMatrix(m2)

if __name__ == '__main__':
    main()

File: cc2.4/cc_2_4.cc
/*
 * rotate matrix 
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int** create_rand_matrix(int row, int col) {
  srand(time(NULL));
  int** ret_ma = (int**) malloc(sizeof(int*) * row);
  for (int i = 0; i < row; i++) {
    ret_ma[i] = (int*) malloc(sizeof(int) * col);
    for (int j = 0; j < col; j++) {
      ret_ma[i][j] = rand() % 10;
    }
  }
  return ret_ma;
}

void delete_matrix(int** ma, int row, int col) {
  if (ma == NULL)
    return;
  for (int i = 0; i < row; i++) {
    if (ma[i] == NULL)
      return;
    free(ma[i]);
  }
  free(ma);
}

void print_matrix(int** ma, int row, int col) {
  for (int i = 0; i < row; i++) {
    for (int j = 0; j < col; j++) {
      printf("%d ", ma[i][j]);
    }
    printf("\n");
  }
}

void rotate_matrix(int** ma, int row, int col) {
  if (row != col)
    return;
  int n = row;
  for (int i = 0; i < n/2; i++) {
    for (int j = 0; j < n - 1 - 2*i; j++) {
      int k = i + j;
      int tmp = ma[i][k];
      ma[i][k] = ma[n-1-k][i];
      ma[n-1-k][i] = ma[n-1-i][n-1-k];
      ma[n-1-i][n-1-k] = ma[k][n-1-i];
      ma[k][n-1-i] = tmp;
    }
  }
}

int main() {
  int** m1 = create_rand_matrix(4, 4);
  print_matrix(m1, 4, 4);
  rotate_matrix(m1, 4, 4);
  printf("\n");
  print_matrix(m1, 4, 4);
  delete_matrix(m1, 4, 4);

  m1 = create_rand_matrix(10, 10);
  print_matrix(m1, 10, 10);
  rotate_matrix(m1, 10, 10);
  printf("\n");
  print_matrix(m1, 10, 10);
  delete_matrix(m1, 10, 10);

  return 0;
}
